src/gui/src/renderThread.cpp

```diff
@@ -979,6 +1045,176 @@ void RenderThread::drawLayer(QPainter* painter,
              layer_timer);
 }
 
+// Draw the region of the block.  Depth is not yet used but
+// is there for hierarchical design support.
+void RenderThread::drawBlockConcurrent(QImage& image,
+                             dbBlock* block,
+                             QVector<QRect>& subRects,
+                             QVector<QFuture<QImage>>& futures,
+                             int threadCount,
+                             const QRect& draw_bounds,
+                             int depth)
+{
+  const Rect dbu_bounds = viewer_->screenToDBU(draw_bounds);
+
+  utl::Timer timer;
+
+  utl::Timer manufacturing_grid_timer;
+  const int instance_limit = viewer_->instanceSizeLimit();
+
+  QPainter painter(&image);
+  painter.setRenderHints(QPainter::Antialiasing);
+
+  GuiPainter gui_painter(&painter,
+                         viewer_->options_,
+                         dbu_bounds,
+                         viewer_->pixels_per_dbu_,
+                         block->getDbUnitsPerMicron());
+
+  // Draw die area, if set
+  painter.setPen(QPen(Qt::gray, 0));
+  painter.setBrush(QBrush());
+  Rect bbox = block->getDieArea();
+  if (bbox.area() > 0) {
+    painter.drawRect(bbox.xMin(), bbox.yMin(), bbox.dx(), bbox.dy());
+  }
+
+  drawManufacturingGrid(&painter, dbu_bounds);
+  debugPrint(logger_,
+             GUI,
+             "draw",
+             1,
+             "manufacturing grid {}",
+             manufacturing_grid_timer);
+
+  /* Search instances */
+  utl::Timer inst_timer;
+  std::vector<dbInst*> insts;
+  std::vector<std::vector<dbInst*>> sub_insts;
+  insts.reserve(10000);
+  {
+    QVector<QFuture<std::vector<dbInst*>>> inst_futures;
+    for (const auto& subRect : subRects) {
+      const Rect dbu_subrect = viewer_->screenToDBU(subRect);
+      inst_futures.push_back(
+          QtConcurrent::run([=]() {
+            auto sub_inst_range = viewer_->search_.searchInsts(block,
+                                            dbu_subrect.xMin(),
+                                            dbu_subrect.yMin(),
+                                            dbu_subrect.xMax(),
+                                            dbu_subrect.yMax(),
+                                            instance_limit);
+            // Cache the search results as we will iterate over the instances
+            // for each layer.
+            std::vector<dbInst*> subInsts;
+            subInsts.reserve(10000);
+            for (auto& [box, inst] : sub_inst_range) {
+              if (restart_) {
+                break;
+              }
+              if (viewer_->options_->isInstanceVisible(inst)) {
+                subInsts.push_back(inst);
+              }
+            }
+            return subInsts;
+          }));
+    }
+    for (auto& future : inst_futures) {
+      future.waitForFinished();
```

Comment:
Is there a need to synchronize here or could the subInsts flow directly into the drawInstanceOutlines from the next loop using then()?